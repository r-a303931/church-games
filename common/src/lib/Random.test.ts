import {
	Random,
	NotRandom,
	SeededRandom,
	getRange,
	getInt,
	getFrom0,
	getIntFrom0,
	shuffleArray,
	generateNumbers,
	uuidv4Options,
	uuid,
	selectRandom,
} from './Random';
import { v4 } from 'uuid';

describe('NotRandom', () => {
	it('should always generate the same number', () => {
		const random: Random = new NotRandom(0.5);

		expect(random.get()).toEqual(0.5);
		expect(random.get()).toEqual(0.5);
		expect(random.get()).toEqual(0.5);
		expect(random.get()).toEqual(0.5);
	});
});

describe('SeededRandom', () => {
	it('should always generate the same numbers with the same seed', () => {
		const random: Random = new SeededRandom(10);

		expect(random.get()).toMatchSnapshot();
		expect(random.get()).toMatchSnapshot();
		expect(random.get()).toMatchSnapshot();
		expect(random.get()).toMatchSnapshot();
	});
});

describe('random library', () => {
	const random: Random = new SeededRandom(10);
	const randomLow: Random = new NotRandom(0);
	const randomHigh: Random = new NotRandom(0.999999999);

	it('should get a number in the range specified', () => {
		expect(getRange(randomHigh)(5)(10)).toBeCloseTo(9.99999999);
		expect(getRange(randomLow)(5)(10)).toEqual(5);
		expect(getRange(random)(5)(10)).toMatchSnapshot();
	});

	it('should get an integer in the range specified', () => {
		expect(getInt(randomHigh)(5)(10)).toEqual(9);
		expect(getInt(randomLow)(5)(10)).toEqual(5);
		expect(getInt(random)(5)(10)).toMatchSnapshot();
	});

	it('should get a number from 0 to the number specified', () => {
		expect(getFrom0(randomHigh)(10)).toBeCloseTo(9.99999999);
		expect(getFrom0(randomLow)(10)).toEqual(0);
		expect(getFrom0(random)(10)).toMatchSnapshot();
	});

	it('should get an integer from 0 to the number specified', () => {
		expect(getIntFrom0(randomHigh)(10)).toEqual(9);
		expect(getIntFrom0(randomLow)(10)).toEqual(0);
		expect(getIntFrom0(random)(10)).toMatchSnapshot();
	});

	it('should shuffle an array', () => {
		const array = [0, 1, 2, 3, 4, 5];

		expect(shuffleArray(randomLow)(array)).toMatchObject(array);
		expect(shuffleArray(randomHigh)(array)).toMatchObject([...array].reverse());
		expect(shuffleArray(random)).toMatchSnapshot();
	});

	it('should generate numbers as an iterator', () => {
		const generatorLow = generateNumbers(randomLow)[Symbol.iterator]();
		const generator = generateNumbers(random)[Symbol.iterator]();
		const generatorHigh = generateNumbers(randomHigh)[Symbol.iterator]();

		expect(generatorLow.next().value).toEqual(0);
		expect(generatorLow.next().value).toEqual(0);
		expect(generatorLow.next().value).toEqual(0);

		expect(generatorHigh.next().value).toBeCloseTo(0.999);
		expect(generatorHigh.next().value).toBeCloseTo(0.999);
		expect(generatorHigh.next().value).toBeCloseTo(0.999);

		expect(generator.next().value).toMatchSnapshot();
		expect(generator.next().value).toMatchSnapshot();
		expect(generator.next().value).toMatchSnapshot();
	});

	it('should generate UUIDs', () => {
		const uuidOptions = uuidv4Options(random);

		expect(v4(uuidOptions)).toMatchSnapshot();
		expect(v4(uuidOptions)).toMatchSnapshot();
		expect(v4(uuidOptions)).toMatchSnapshot();

		expect(uuid(random)()).toMatchSnapshot();
		expect(uuid(random)()).toMatchSnapshot();
		expect(uuid(random)()).toMatchSnapshot();
	});

	it('should select an element from an array', () => {
		const elems = [0, 1, 2, 3];

		expect(selectRandom(randomLow)(elems)).toEqual(elems[0]);
	});
});
