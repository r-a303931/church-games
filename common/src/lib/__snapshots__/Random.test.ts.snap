// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeededRandom should always generate the same numbers with the same seed 1`] = `0.7888911063018895`;

exports[`SeededRandom should always generate the same numbers with the same seed 2`] = `0.0979344929655781`;

exports[`SeededRandom should always generate the same numbers with the same seed 3`] = `0.27081999565052683`;

exports[`SeededRandom should always generate the same numbers with the same seed 4`] = `0.6703682664092412`;

exports[`random library should generate UUIDs 1`] = `"0620c573-8e7c-4b37-bb95-c20ea9af9f44"`;

exports[`random library should generate UUIDs 2`] = `"d32c369e-aff4-41c5-8840-0409e2bb7479"`;

exports[`random library should generate UUIDs 3`] = `"4a46401c-727d-45b9-a1e4-d2318e42497d"`;

exports[`random library should generate UUIDs 4`] = `"b926e88b-3b47-492f-9333-c8df1d1e49a5"`;

exports[`random library should generate UUIDs 5`] = `"3e6ad2fb-b1f7-4056-ad7a-9de0132eee57"`;

exports[`random library should generate UUIDs 6`] = `"44e2c6a5-93d1-4f6d-927c-0b2ecd5a9d05"`;

exports[`random library should generate numbers as an iterator 1`] = `0.07355694870420848`;

exports[`random library should generate numbers as an iterator 2`] = `0.8784015711680695`;

exports[`random library should generate numbers as an iterator 3`] = `0.9668333493468708`;

exports[`random library should get a number from 0 to the number specified 1`] = `2.7081999565052683`;

exports[`random library should get a number in the range specified 1`] = `8.944455531509448`;

exports[`random library should get an integer from 0 to the number specified 1`] = `6`;

exports[`random library should get an integer in the range specified 1`] = `5`;

exports[`random library should shuffle an array 1`] = `[Function]`;
